python/lsst/meas/deblender/baseline.py

```diff
@@ -393,6 +395,121 @@ def deblend(footprint, maskedImage, psf, psffwhm,
+                if pkres.skip:
+                    continue
+                pkres.templateImage *= X1[index]
+                pkres.setTemplateWeight(X1[index])
+                index += 1
+
+
+        exitLoop = True
+
+        # If galaxies have substructure, such as face-on spirals, the process of identifying peaks can
+        # "shred" the galaxy into many pieces.  The templates of shredded galaxies are typically quite
+        # similiar because they represent the same galaxy.  We try to identify these "degenerate" peaks
+        # by looking at the inner product (in pixel space) of pairs of templates.  If they are nearly
+        # parallel, we only keep one of the peaks an reject the other.
+        if removeDegenerateTemplates:
+
+            log.trace('Looking for degnerate templates')
+
+            nchild = np.sum([pkres.skip is False for pkres in res.peaks])
+            indexes = [pkres.pki for pkres in res.peaks if pkres.skip is False]
+
+            # We build a matrix that stores the dot product between templates.
+            # We convert the template images to HeavyFootprints because they already have a method
+            # to compute the dot product.
+            A = np.zeros((nchild, nchild))
+            maxTemplate = []
+            heavies = []
+            for pkres in res.peaks:
+                if pkres.skip:
+                    continue
+                heavies.append(afwDet.makeHeavyFootprint(pkres.templateFootprint,
+                                                         afwImage.MaskedImageF(pkres.templateImage)))
+                maxTemplate.append(np.max(pkres.templateImage.getArray()))
+
+            for i in range(nchild):
+                for j in range(i+1):
+                    A[i, j] = heavies[i].dot(heavies[j])
+
+            # Normalize the dot products to get the cosine of the angle between templates
+            for i in range(nchild):
+                for j in range(i):
+                    norm = A[i, i]*A[j, j]
+                    if norm <= 0:
+                        A[i, j] = 0
+                    else:
+                        A[i, j] /= np.sqrt(norm)
+
+            # Iterate over pairs of objects and find the maximum non-diagonal element of the matrix.  Exit the loop once we
+            # find a single degenerate pair greater than the threshold.
+            rejectedIndex = -1
+            foundReject = False
+            for i in range(nchild):
+                currentMax = 0.
+                for j in range(i):
+                    if A[i, j] > currentMax:
+                        currentMax = A[i, j]
+                        if currentMax > maxTempDotProd:
+                            foundReject = True
+                            rejectedIndex = j
+
+                if foundReject:
+                    break
+
+            del A
+
+            # If one of the objects is identified as a PSF keep the other one, otherwise keep the one
+            # with the maximum template value
```

Comment:
As you say, this comes straight from SDSS:
```
/*
 * If one of a pair of templates has to be deleted, and if only one is
 * DEBLENDED_AS_PSF, choose that one, otherwise choose the one with
 * the _smaller_ peak value in the associated child.
 */
```

I don't remember exactly why I did this like this.  It says that if we have similar templates prefer the non-PSF one;  that seems safer (as using a PSF for a non-PSF object leaves the wings behind) and maybe that was my logic?  Sorry.